airbyte-webapp/src/components/NetworkErrorBoundary/NetworkErrorBoundary.tsx

```diff
@@ -14,9 +14,9 @@ class NetworkErrorBoundary extends React.Component<
   static getDerivedStateFromError(error: {
     message: string;
     status?: number;
-  }): { unReachServer: boolean } {
+  }): { unReachServer: boolean, status: number } {
```

Comment:
I thought this section needed to read more like:

```
  static getDerivedStateFromError(error: {
    message: string;
    status?: number;
  }): { unReachServer: boolean } {
    // Update state so the next render will show the fallback UI.
    return {
      unReachServer:
        error.message === "Failed to fetch" ||
        (error.status !== undefined && error.status === 502),
    };
  }
```

It's trying to set the `unReachServer` boolean to whether or not it should show the "server is loading error". We want to show that error when there's an error message that contains "Failed to fetch" or if the status code is set and the status code is 502 (which means the underlying service we're proxying through isn't available yet).

However, both versions (the one in this PR and the version above), I got a `Uncaught SyntaxError: Unexpected token < in JSON at position 0`. I believe this is because `/api/v1/workspaces/get` is getting an HTML `An error occurred` 502 page from nginx, which is not JSON.

@Jamakase what's the best way to handle this?

Reply:
In PR #3438 I totally refactored how we handle errors and introduced a richer error hierarchy. You could merge it as is and I will update it to the same format.

It also makes parsing error's response optional, so it won't fail if the error returned is not json.

Reply:
Oh that's great.

I think that since this won't actually show the "server is loading" error until  #3438 it probably makes sense to wait until  #3438  is merged, merge the changes into this PR and then merge the PR? Otherwise first time users will be confused about the behavior like before we added the "server is loading" error in the first place.

Reply:
@jrhizor sounds reasonable to me. I need to address few comments for that PR and after that we could merge. Should I merge master into this branch as soon as we are ready for merging?

Reply:
Yep, that'd be perfect.